ships/0024-surfacing-errors-to-buildrun.md

```diff
@@ -0,0 +1,187 @@
+`stdout`. 
+For a better user experience, we propose that a strategy
+should be able to present the failure reason in a compact way that can
+be represented in a BuildRun CRD, for example, as a TaskRun result. This
+task is more straightforward since a TaskRun is aware of its tooling
+and possible errors, which it can emit similarly to [digest](https://github.com/shipwright-io/build/blob/main/samples/buildstrategy/buildkit/buildstrategy_buildkit_cr.yaml#L65).
+
+Additionally, the proposal enables a user to keep the troubleshooting
+process on the BuildRun level and only descend further into Tekton
+specifics when necessary.
+
+### Goals
+
+- A mechanism to automatically pick up error results from a TaskRun
+  Surface them to the associated BuildRun
+
+### Non-Goals
+
+- The specifics of errors in regards to tooling are not part of this
+  proposal. Neither is their means for reporting them. Errors highly
+  depend on each strategy
+
+## Proposal
+
+We propose to introduce two generic results for errors:
+`shp-error-message` and `shp-error-reason`.  The result
+`shp-error-reason` have a recognizable name that is ideally prefixed
+with the problematic technology, for example, `git` in case of missing
+a secret. The result `shp-error-message` should be descriptive text to
+give the user enough information to troubleshoot the BuildRun.  A task
+run emits the new results as Tekton results.  The BuildRun CRD is
+assumed to have the `status.results` according to `0023` (alternatives
+are discussed further below). 
+Finally, we must extend the BuildRun controller
+logic to propagate the new results into the BuildRun CRD's
+`status.results` property.
+
+BuildRun with new error results: 
+
+```yaml
+status:
+   results:
+       - name: shp-error-reason
+         value: git-ssh-auth
+       - name: shp-error-message
+         value: Authentication failed for ...
+```
+
+We do not suggest to mutate the `status.conditions.message` field
+since it contains valuable hints to troubleshoot on a lower level if
+the propagated error message and reason is not sufficient.
+
+### User Stories
+
+#### Story 1 - Developer and direct user of shipwright perspective
+As a user of Shipwright and developer, I want to identify why my
+BuildRun is in a failed state without going through the lengthy
+process of identifying the correct container and manually processing
+its logs.
+#### Story 2 - Github Actions developer (third party application developer)
+As a third-party application developer that uses shipwright under the hood, I want to supply my users with meaningful hints to fix their builds, thus requiring a straightforward API to consume error details.
+
+### Implementation Notes
+
+`tbd in the next phase of the proposal`
+
+### Test Plan
+
+`tbd in the next phase of the proposal`
+
+### Release Criteria
+
+`tbd in the next phase of the proposal`
+
+### Risks and Mitigations
+Tekton results are limited in size (4096 bytes) which applies to the
+proposed error results.
```

Comment:
In general, I don't see a mitigation for the size limit. We can make sure that we cap and validate the length of our error messages to prevent failing due to too large results (`Termination message is above max allowed size 4096, caused by large task result`).

The proposed alternative in the tekton docs of using workspaces should not be applicable for our usage of tekton.

Reply:
Yeah, well, we should still document that there is a limit. Otherwise strategy writers might try to put whole error logs or stack traces into the error message and that should not happen.

Reply:
Ok I miss understood your comment. I have added documentation as mitigation.